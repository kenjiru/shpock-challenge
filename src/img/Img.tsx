import * as React from "react";
import { Component, ReactElement } from "react";

class Img extends Component<IImgProps> {
    public render(): ReactElement<HTMLElement> {
        return (
            <img
                className={this.props.className}
                src={`img/${this.props.src}`}
            />
        );
    }
}

interface IImgProps {
    className?: string;
    src: string;
}

export default Img;
