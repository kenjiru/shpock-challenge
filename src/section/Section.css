.section .section-header {
    display: flex;
    flex-direction: row;
    margin-bottom: 4px;
}

.section .section-header .section-title {
    font-weight: bold;
    flex-grow: 2;
    line-height: 24px;
}

.section .section-header .section-icon {
    width: 24px;
    height: 24px;
    margin-right: 8px;
}

.section .section-content {
    margin: 8px 0 8px 32px;
}