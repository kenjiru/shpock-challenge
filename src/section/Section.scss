.section {
    .section-header {
        display: flex;
        flex-direction: row;
        margin-bottom: 4px;

        .icon-container {
            display: flex;
            flex-direction: column;
            align-items: center;
            justify-content: center;
        }

        .text {
            flex-grow: 2;

            .title {
                font-weight: bold;
                line-height: 24px;
            }

            .sub-title {
                font-size: 12px;
                color: #848484;
            }
        }

        .icon {
            width: 24px;
            height: 24px;
            margin-right: 8px;
        }
    }

    .section-content {
        margin: 8px 0 8px 32px;
    }
}
