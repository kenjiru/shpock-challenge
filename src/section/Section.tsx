import * as classNames from "classnames";
import * as React from "react";
import { Component, ReactElement } from "react";

import "./Section.css";

class Section extends Component<ISectionProps> {
    public render(): ReactElement<HTMLElement> {
        return (
            <div className={this.getClassName()}>
                <div className="section-header">
                    <img
                        className="section-icon"
                        src={"img/" + this.props.icon}
                    />
                    <div className="section-title">{this.props.title}</div>
                    {this.renderTitleRight()}
                </div>
                <div className="section-content">
                    {this.props.children}
                </div>
            </div>
        );
    }

    private renderTitleRight(): string | ReactElement<HTMLElement> {
        if (typeof this.props.titleRight === "undefined") {
            return <div/>;
        }

        return (
            <div className="section-right">
                {this.props.titleRight}
            </div>
        );

    }

    private getClassName(): string {
        return classNames("section", this.props.className);
    }
}

interface ISectionProps {
    icon: string;
    title: string;
    titleRight?: string | ReactElement<HTMLElement>;
    className?: string;
}

export default Section;
