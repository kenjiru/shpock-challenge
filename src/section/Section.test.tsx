import * as React from "react";
import { shallow, ShallowWrapper } from "enzyme";
import shallowToJson from "enzyme-to-json";

import Section from "./Section";
import Img from "../img/Img";

describe("Section", () => {
    it("should render correctly", () => {
        const output = shallow(
            <Section className="custom-class"/>
        );
        expect(shallowToJson(output)).toMatchSnapshot();
    });

    it("should render an icon if an icon prop is present", () => {
        const ICON: string = "foo.png";
        const output: ShallowWrapper = shallow(
            <Section icon={ICON}/>
        );

        expect(output.contains(<Img className="icon" src={ICON}/>)).toBe(true);
    });

    it("should NOT render an icon if an icon prop is NOT present", () => {
        const output: ShallowWrapper = shallow(
            <Section/>
        );

        expect(output.find(Img).length).toBe(0);
    });

    it("should render the title right if the prop is present", () => {
        const titleRight: string = "Title Right";
        const output: ShallowWrapper = shallow(
            <Section titleRight={titleRight}/>
        );

        expect(output.contains(<div className="title-right">{titleRight}</div>)).toBe(true);
    });

    it("should handle clicking on the header", () => {
        const handleClick = jest.fn();
        const output: ShallowWrapper = shallow(
            <Section onHeaderClick={handleClick}/>
        );

        output.find(".section-header").simulate('click');

        expect(handleClick).toHaveBeenCalled();
    });
});
