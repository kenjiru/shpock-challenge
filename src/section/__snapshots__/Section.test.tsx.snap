// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Section should render correctly 1`] = `
<div
  className="section custom-class"
>
  <div
    className="section-header"
  >
    <div
      className="icon-container"
    >
      <div
        className="icon"
      />
    </div>
    <div
      className="text"
    >
      <div
        className="title"
      />
      <div
        className="sub-title"
      />
    </div>
    <div />
  </div>
  <div
    className="section-content"
  />
</div>
`;
